### PR TITLE
Fix data classes

### PR DESCRIPTION
**Pull request by Xaivteev**
JobSpecification, JobCompletionData, SimulationSpecification, Task, and Machine are just data classes.

To fix this, we would merge the data classes with the classes that access them, or pull the methods from classes that access them if there are multiple, depending on the implementation. Specifically, we've moved Machine into MachineShopSimulator, Task into Job, JobSpecification into SimulationSpecifcation, and JobCompletionData into SimulationResults. We also removed dependencies on these data classes in the tests and in other classes. This was done by writing new methods in the classes they were made a part of. In this way, there are fewer inter-dependencies between classes. The newly created inner classes are also more secure due to being made private.

### CODE REVIEW

File: src/applications/MachineShopSimulator.java [hunk 1]
```diff
@@ -1,26 +1,28 @@
-/** machine shop simulation */
+/** machineArray shop simulation */
```
Comment (rabex028) [inline src/applications/MachineShopSimulator.java:1]: Was the change to this comment intended?

File: src/applications/MachineShopSimulator.java [hunk 2]
```diff
@@ -45,72 +47,75 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
         }
     }
 
     private static void setUpJobs(SimulationSpecification specification) {
         // input the jobs
         Job theJob;
         for (int i = 1; i <= specification.getNumJobs(); i++) {
-            int tasks = specification.getJobSpecifications(i).getNumTasks();
-            int firstMachine = 0; // machine for first task
+//            int tasks = specification.getJobSpecifications(i).getNumTasks();
```
Comment (rabex028) [inline src/applications/MachineShopSimulator.java]: This line might be unnecessary

**Review by YujingSong — APPROVED** [review]
Good job!  Renaming provides more readability especially in machineArray.   It's kind too much for MachineShopSimulator to take Machine as its private class, I mean, maybe you can still maintain Machine as a independent class, but we can make it only touch MachineShopSimulator.  Another advice is that, that will be great if you can delete all the useless code, which already been commented, otherwise it is easy to be misunderstood.

File: src/applications/SimulationResults.java [hunk 3]
```diff
@@ -68,4 +68,49 @@ public void setJobCompletionData(int jobNumber, int completionTime, int totalWai
...
+
+        public int getTotalWaitTime() {
+            return totalWaitTime;
+        }
+
+        public int getJobNumber() {
+            return jobNumber;
+        }
+    }
 }
```
Comment (holma198) [inline src/applications/SimulationResults.java:116]: Definitely an improvement over having that data class, and still retaining functionality

**Review by holma198 — CHANGES_REQUESTED** [review]
### Code Review - RJ Holman

 ##### Documentation
The pull request does a decent job of explaining what was done.  It could specify more on why these changes were made, but the request is overall sufficient.  The smell is clearly Data classes.

 ##### Commits
The commits are just frequent enough to not set off any red flags, however, in the future I would do them even more frequently.  For instance removing a class such as in [830b757b6bea78d504d6f3076de2e974d9511eea], this could really be two separate commits, one for moving the functionality, and one for removing the class.
 ##### refactoring quality
There is definitely an improvement here.  The superfluous data classes have been removed to a point where I would claim that the data class issue has been closed.  There is also commits dedicated to renaming variables and cleaning up changes made.  Improving the code as you refactor is a part of refactoring itself, so these commits are certainly helpful and reflect that the code had been successfully refactored with the detail necessary to do so.  The only issue I can see is that commit #7 includes xml and iml files that shouldn’t be updated.
 #### Bottom Line
 - remove `iml` and `xml` changes from [6af03d3628016045dcff40ae02a18d00bb604c48]
 - that’s all I request, then this is ready to merge

File: src/applications/SimulationSpecification.java [hunk 4]
```diff
@@ -35,13 +36,40 @@ public int getChangeOverTimes(int machineNumber) {
...
+//        this.jobSpecifications = jobSpecifications;
+//    }
+//
+//    public void setJobSpecification(int tasksNumber, int[] tasks) {
+//        this.jobSpecifications = jobSpecifications;
+//    }
+//
+//    public JobSpecification getJobSpecifications(int jobNumber) {
+//        return jobSpecifications[jobNumber];
+//    }
```
Comment (CMenne) [inline src/applications/SimulationSpecification.java]: Is this necessary to have still?

**Review by CMenne — APPROVED** [review]
I think that your issues could be documented better. However, it looks like they were recently created. Our group forgot to add details as we thought they were covered in the smells report. In the next round of refactoring we plan to better document the issues. Your pull request looks well documented and explains what you've done and why very well. :+1: One thing that I recommend for the future is to make your pull requests a bit smaller. One of the pull request readings for class mentions that they would probably cap there pull requests at 200 lines of code. Their reasoning being that they have a 90% chance of completing the review within an hour. One suggestion would to have created a pull request for each case of data class that you've fixed. However, what you've done looks to have cleaned up Data Class smell and remove those files. One thing that I think you guys could address is the amount of commented out code in these commits.

**Comment by rabex028** [comment]
Overall I liked the changes your team made to the code base, especially with the name
refactorings you've made. They do a really good job of making things readable.
I think using nested classes to fix the data class issue is also a good solution and
makes things more compact (and the groupings make sense). There are a good number of commits
with informative titles. It looks like you've done a substantial amount of work.

There are several things I might consider addressing before pulling these changes into master however. For one, it looks like there are large amounts of old commented out code that I'm not sure is needed to be kept anymore that would be merged in with the pull request. Another issue is that some actual code
documentation has been altered in ways that don't seem to make sense anymore. In particular, comments that now reference machine arrays instead of machines do not always make as much sense. Addressing these issues in addition to removing the unnecessary iml files would make these commits a great addition to master :+1:

**Review by rabex028 — APPROVED** [review]
Overall I liked the changes your team made to the code base, especially with the name
refactorings you've made. They do a really good job of making things readable.
I think using nested classes to fix the data class issue is also a good solution and
makes things more compact (and the groupings make sense). There are a good number of commits
with informative titles. It looks like you've done a substantial amount of work.

There are several things I might consider addressing before pulling these changes into master however. For one, it looks like there are large amounts of old commented out code that I'm not sure is needed to be kept anymore that would be merged in with the pull request. Another issue is that some actual code
documentation has been altered in ways that don't seem to make sense anymore. In particular, comments that now reference machine arrays instead of machines do not always make as much sense. Addressing these issues in addition to removing the unnecessary iml files would make these commits a great addition to master :+1:

File: src/applications/SimulationSpecification.java (hunk 4 above)
Comment (nickplucker) [reply src/applications/SimulationSpecification.java]: I agree with Chuck. Maybe this should be removed?

File: tests/properties/applications/SimulationProperties.java [hunk 5]
```diff
@@ -54,11 +55,15 @@ public void jobsOutputInTimeOrder(
     {
         final SimulationResults results = MachineShopSimulator.runSimulation(specification);
 
-        JobCompletionData[] jobCompletionData = results.getJobCompletionData();
-        for (int i=1; i<jobCompletionData.length-1; ++i) {
-            assertThat(jobCompletionData[i].getCompletionTime(),
-                    lessThanOrEqualTo(jobCompletionData[i+1].getCompletionTime()));
-        }
+//        SimulationResults.JobCompletionData[] jobCompletionData = results.getJobCompletionData();
```
Comment (nickplucker) [inline tests/properties/applications/SimulationProperties.java:58]: Perhaps you want to get rid of this unused code?

**Review by nickplucker — APPROVED** [review]
Overall, you did a nice job squashing the data classes in the code base. The code makes a lot more sense and the amount of classes has been reduced. 

### Some suggestions:
- I would commit more often. You do a decent job committing after a change has been made, but there are a few commits that feel a tad large [6d0f32d3ac6d9b004ace33997cd7357f4dacbacd]. 
- One of the readings that we were assigned suggested that pull requests shouldn't be too large (~200 lines of code) and cover a broad range of items. While this pull request revolves around data classes, perhaps it would have been better to make multiple pull requests for each data class you eliminated. This would make each pull request be more specific and easier to review. 
- As you refactored, I noticed that there was quite a few lines/blocks of commented out code. To improve the readability, I would suggest deleting those lines if they are not needed anymore.
- As suggested above, some of the comments don't make sense anymore. As the code was changed, so was the function of it. Spend some time changing the comments so they make more sense. 
- As with many groups, remove the `.iml` files. 

I think that what you did was a major improvement to the code and this would be a great addition to `master`. Well done!

**Comment by miyat001** [comment]
Your pull request is well documented. You explained what you changed. 
Your commit is little short, so you should explain what you changed about code.
Rename is small change, but it improve the readability of code.
To Combine "getFirstTask()" and "getFirstMachine()" is great change. It makes code shorter and reduce the function call.
I think Machine Shop Simulator class is very long. This class has 246 lines. To move Machine class to this class is good idea, but you should move another function to another class. For example, change-state function.

**Review by miyat001 — APPROVED** [review]
